File: pretty.py
"""
Generic Pretty Printer based on:

Pretty Priting
Derek C. Oppen, 1979
https://www.cs.tufts.edu/~nr/cs257/archive/derek-oppen/prettyprinting.pdf


This version does away with all the space optimizations and
offline concerns of the paper resulting in a drastically
simpler algorithm.

The API consists of a single function `PrettyPrint`.
Which takes list of Tokens and returns a string with
formatted text.

A token may be one of:
* String
* Break
* Begin
* End

A new BreakType, FORCE_LINE_BREAK, is added to make nesting
easier.

For usage see `tests.py`


Copyright 2025 Robert Muth <robert@muth.org>
See LICENSE for more details.
"""
from typing import Union

import enum
import dataclasses


class BreakType(enum.Enum):
    # No breaks are needed (the block fits on the line).
    # This will be automatically selected.
    FITS = 1
    # Fill line as much as possible then issue a line break and
    # continue in the same fashion on the next line
    INCONSISTENT = 2
    # If a block does not fit on a line, put each subblock
    # on a new line.
    CONSISTENT = 3
    # NOTE: THIS IS AN ADDITION TO THE ORIGINAL PAPER
    # Put each subblock on a new line
    # The `offset` applies even to the first line.
    # So this can be used as a nesting construct.
    FORCE_LINE_BREAK = 4


@dataclasses.dataclass()
class String:
    string: str


@dataclasses.dataclass()
class Break:
    """
    Either a space of widths `num_spaces` or a line break with additional
    `offset` relative to the indentation of the enclosing block
    """
    num_spaces: int = 1  # number of spaces if no overflow
    offset: int = 0  # indent for overﬂow lines


def LineBreak(offset=0):
    # Break which is guaranteed to overflow
    return Break(num_spaces=1000000, offset=offset)


@dataclasses.dataclass()
class Begin:
    # how to handle Breaks nested between this `Begin`` and `End``
    break_type: BreakType
    # additional indent
    offset: int


@dataclasses.dataclass()
class End:
    pass


Token = Union[String, Break, Begin, End]


def _ComputeSizes(tokens: list[Token]):
    # Stack invariants:
    # only Begin and Break tokens can be on the stack
    # there are never two consecutive Break Tokens on the Stack
    # So the top of the stack is eiher [... Begin] or
    # [... Begin Break]
    scan_stack = []
    sizes = []
    total = 0  # corresponds to `rightotal` in the paper
    for n, t in enumerate(tokens):
        if isinstance(t, Begin):
            scan_stack.append(n)
            sizes.append(-total)
        elif isinstance(t, End):
            sizes.append(1)
            x = scan_stack.pop(-1)
            sizes[x] += total
            if isinstance(tokens[x], Break):
                x = scan_stack.pop(-1)
                sizes[x] += total
        elif isinstance(t, Break):
            sizes.append(-total)
            z = scan_stack[-1]
            if isinstance(tokens[z], Break):
                z = scan_stack.pop()
                sizes[z] += total
            scan_stack.append(n)
            total += t.num_spaces
        elif isinstance(t, String):
            sizes.append(len(t.string))
            total += len(t.string)
        else:
            assert False, f"unknown token {t}"
    assert len(tokens) == len(sizes), f"{len(tokens)} {len(sizes)}"
    return sizes


class _Output:

    def __init__(self: "_Output", line_width: int):
        self.buffer: list[str] = []
        self.remaining = line_width  # how much space is left in current line
        self.line_width = line_width  # constant

    def append(self, s):
        self.buffer.append(s)

    def get_string(self):
        return "".join(self.buffer)

    def append_with_space_update(self, s):
        self.buffer.append(s)
        self.remaining -= len(s)

    def current_indent(self):
        return self.line_width - self.remaining

    def fits_in_current_line(self, size: int) -> bool:
        return size <= self.remaining

    def set_offset(self, offset: int):
        self.remaining = offset

    def line_break(self):
        self.buffer.append("\n")
        self.buffer.append(" " * self.current_indent())


@dataclasses.dataclass()
class _Entry:
    offset: int  # same measurement as `remaining`, i.e. no offset means `offset`== margin
    brk: BreakType


def _Render(tokens, sizes, output: _Output):
    print_stack: list[_Entry] = []

    for token, size in zip(tokens, sizes):
        if isinstance(token, Begin):
            if token.break_type == BreakType.FORCE_LINE_BREAK:
                if print_stack:
                    offset = print_stack[-1].offset
                    output.set_offset(offset - token.offset)
                    output.line_break()
                else:
                    offset = output.line_width
                entry = _Entry(
                    output.remaining, token.break_type)
            elif output.fits_in_current_line(size):
                entry = _Entry(0, BreakType.FITS)
            else:
                bt = (BreakType.CONSISTENT
                      if token.break_type == BreakType.CONSISTENT
                      else BreakType.INCONSISTENT)
                entry = _Entry(
                    output.remaining - token.offset, bt)
            print_stack.append(entry)
        elif isinstance(token, End):
            top: _Entry = print_stack.pop()
            if top.brk == BreakType.FORCE_LINE_BREAK:
                offset = output.line_width if not print_stack else print_stack[-1].offset
                output.set_offset(offset)
                output.line_break()
            # Note, not resetting the offset right away in the
            # non-FORCE_LINE_BREAK case is responsible for the
            # behavior characterized in the paper as:
            # "In particular, it effectively changes (dynamically)
            # each occurrence of ] (string) into (string) ]."
        elif isinstance(token, Break):
            top = print_stack[-1]
            if top.brk == BreakType.FITS:
                output.append_with_space_update(
                    " " * token.num_spaces)  # indent
            elif top.brk in (BreakType.CONSISTENT, BreakType.FORCE_LINE_BREAK):
                output.set_offset(top.offset - token.offset)
                output.line_break()
            elif top.brk == BreakType.INCONSISTENT:
                if output.fits_in_current_line(size):
                    output.append_with_space_update(
                        " " * token.num_spaces)  # indent
                else:
                    output.set_offset(top.offset - token.offset)
                    output.line_break()

        elif isinstance(token, String):
            # TODO: handle long/multiline strings
            assert size == len(token.string)
            output.append_with_space_update(token.string)
        else:
            assert False, f"Unsupported token {token}"


def PrettyPrint(tokens: list[Token], line_width: int) -> str:
    # print(tokens)
    output = _Output(line_width)
    sizes = _ComputeSizes(tokens)
    _Render(tokens, sizes, output)
    return output.get_string()

File: pretty.cc
// (c) Robert Muth - see LICENSE for more info

#include "pretty.h"

#include <cassert>

namespace PP {

std::vector<ssize_t> ComputeSizes(const std::vector<Token>& tokens) {
  std::vector<ssize_t> sizes;
  sizes.reserve(tokens.size());
  ssize_t total = 0;
  std::vector<size_t> scan_stack;
  size_t x;

  for (size_t i = 0; i < tokens.size(); i++) {
    const Token& token = tokens[i];
    switch (token.type) {
      case TokenType::BEG:
        scan_stack.push_back(i);
        sizes.push_back(-total);
        break;
      case TokenType::END:
        sizes.push_back(1);
        x = scan_stack.back();
        sizes[x] += total;
        scan_stack.pop_back();
        if (tokens[x].type == TokenType::BRK) {
          x = scan_stack.back();
          sizes[x] += total;
          scan_stack.pop_back();
        }
        break;
      case TokenType::BRK:
        sizes.push_back(-total);
        // maybe close sizes a preceeding BRK
        x = scan_stack.back();
        if (tokens[x].type == TokenType::BRK) {
          scan_stack.pop_back();
          sizes[x] += total;
        }
        scan_stack.push_back(i);
        total += token.brk.num_spaces;
        break;
      case TokenType::STR:
        total += token.str.size();
        break;
      default:
        assert(false);  // unreachable
        break;
    }
    sizes.push_back(total);
  }
  return sizes;
}

void UpdatesSizesForNoBreaks(const std::vector<Token>& tokens,
                             std::vector<ssize_t>& sizes) {
  ssize_t total = INFINITE_WIDTH;
  for (size_t j = 0; j < tokens.size(); j++) {
    size_t i = tokens.size() - 1 - j;
    const Token& token = tokens[i];
    switch (token.type) {
      case TokenType::BEG:
        if (token.beg.break_type == BreakType::FORCE_LINE_BREAK) {
          total = INFINITE_WIDTH;
        }
        break;
      case TokenType::END:
        total += INFINITE_WIDTH;
        break;

      case TokenType::BRK:
        if (token.brk.nobreak) {
          if (total < sizes[i]) {
            sizes[i] = total;
            total += token.brk.num_spaces;
          } else {
            total = sizes[i];
          }
        } else {
          total = 0;
        }
        break;
      case TokenType::STR:
        total += token.str.size();
        break;
      default:
        assert(false);  // unreachable
        break;
    }
  }
  //
  total = 0;
  for (size_t j = 0; j < tokens.size(); j++) {
    size_t i = tokens.size() - 1 - j;
    const Token& token = tokens[i];
    switch (token.type) {
      case TokenType::BEG:
        if (token.beg.break_type == BreakType::FORCE_LINE_BREAK) {
          total = 0;
        }
        break;
      case TokenType::END:
        //
        break;

      case TokenType::BRK:
        total += token.brk.num_spaces;
        if (!token.brk.nobreak) {
          if (total > sizes[i]) {
            sizes[i] = total;
          } else {
            total = 0;
          }
        }
        break;
      case TokenType::STR:
        total += token.str.size();
        break;
      default:
        assert(false);  // unreachable
        break;
    }
  }
}

class Output {
 public:
  Output(size_t line_width, size_t approx_output_length)
      : line_width_(line_width), remaining_(line_width) {
    buffer_.reserve(approx_output_length);
  }

  void Append(std::string_view str) { buffer_.append(str); }

  void AppendWithSpaceUpdate(std::string_view str) {
    buffer_ += str;
    remaining_ -= str.size();
  }

  void IndentWithSpaceUpdate(size_t num_spaces) {
    for (size_t i = 0; i < num_spaces; i++) {
      buffer_ += ' ';
    }
    remaining_ -= num_spaces;
  }

  size_t CurrentIndent() { return line_width_ - remaining_; }

  size_t LineWidth() { return line_width_; }

  size_t Remaining() { return remaining_; }

  bool FitsInCurrentLine(size_t size) { return size <= remaining_; }

  std::string Get() { return buffer_; }

  void SetOffsetAndLineBreak(size_t offset) {
    remaining_ = offset;
    buffer_ += '\n';
    size_t ci = CurrentIndent();
    for (size_t i = 0; i < ci; i++) {
      buffer_ += ' ';
    }
  }

 private:
  size_t line_width_;
  size_t remaining_;
  std::string buffer_;
};

struct Entry {
  size_t offset;
  BreakType break_type;
};

void Render(const std::vector<Token>& tokens, const std::vector<ssize_t>& sizes,
            Output* output) {
  std::vector<Entry> stack;
  for (size_t i = 0; i < tokens.size(); i++) {
    const Token& token = tokens[i];
    const size_t size = sizes[i];
    switch (token.type) {
      case TokenType::BEG: {
        Entry entry;
        if (token.beg.break_type == BreakType::FORCE_LINE_BREAK) {
          size_t offset;
          if (!stack.empty()) {
            offset = stack.back().offset;
            output->SetOffsetAndLineBreak(offset);
          } else {
            offset = output->LineWidth();
          }
        } else if (output->FitsInCurrentLine(size)) {
          entry = {0, BreakType::FITS};
        } else {
          entry = {
              output->Remaining(),
              token.beg.break_type == BreakType::CONSISTENT
                  ? BreakType::CONSISTENT
                  : BreakType::INCONSISTENT,
          };
        }
        stack.push_back(entry);
      } break;
      case TokenType::END:
        stack.pop_back();
        break;
      case TokenType::BRK: {
        const Entry& top = stack.back();
        BreakType break_type = top.break_type;
        size_t offset = top.offset;
        stack.pop_back();
        if ((token.brk.nobreak && output->FitsInCurrentLine(size)) ||
            break_type == BreakType::FITS) {
          output->IndentWithSpaceUpdate(token.brk.num_spaces);
        } else if (top.break_type == BreakType::CONSISTENT ||
                   top.break_type == BreakType::FORCE_LINE_BREAK) {
          output->SetOffsetAndLineBreak(offset - token.brk.offset);
        } else if (top.break_type == BreakType::INCONSISTENT) {
          if (token.brk.nobreak && output->FitsInCurrentLine(size)) {
            output->IndentWithSpaceUpdate(token.brk.num_spaces);
          } else {
            output->SetOffsetAndLineBreak(offset - token.brk.offset);
          }
        }
      } break;

      case TokenType::STR:
        output->AppendWithSpaceUpdate(token.str);
        break;
      case TokenType::INVALID:
        assert(false);  // unreachable
        break;
    }
  }
}

size_t ApproxOutputLength(const std::vector<Token>& tokens) {
  size_t total = 0;
  for (const Token& token : tokens) {
    switch (token.type) {
      case TokenType::BEG:
      case TokenType::END:
        break;
      case TokenType::STR:
        total += token.str.size();
        break;
      case TokenType::BRK:
        total += token.brk.num_spaces;
        break;
      default:
        assert(false);  // unreachable
        break;
    }
  }
  return total;
}

std::string PrettyPrint(const std::vector<Token>& tokens, size_t line_width) {
  std::vector<ssize_t> sizes = ComputeSizes(tokens);
  UpdatesSizesForNoBreaks(tokens, sizes);
  Output output(line_width, ApproxOutputLength(tokens));
  Render(tokens, sizes, &output);
  return output.Get();
}

}  // namespace PP
